bundles/org.openhab.binding.radiothermostat/src/main/java/org/openhab/binding/radiothermostat/internal/handler/RadioThermostatHandler.java

```diff
@@ -0,0 +1,600 @@
+        if (isLinked(channelId)) {
+            Object value;
+            try {
+                value = getValue(channelId, rthermData);
+            } catch (Exception e) {
+                logger.debug("Error setting {} value", channelId.toUpperCase());
+                return;
+            }
+
+            State state = null;
+            if (value == null) {
+                state = UnDefType.UNDEF;
+            } else if (value instanceof PointType) {
+                state = (PointType) value;
+            } else if (value instanceof ZonedDateTime) {
+                state = new DateTimeType((ZonedDateTime) value);
+            } else if (value instanceof QuantityType<?>) {
+                state = (QuantityType<?>) value;
+            } else if (value instanceof BigDecimal) {
+                state = new DecimalType((BigDecimal) value);
+            } else if (value instanceof Integer) {
+                state = new DecimalType(BigDecimal.valueOf(((Integer) value).longValue()));
+            } else if (value instanceof String) {
+                state = new StringType(value.toString());
+            } else if (value instanceof OnOffType) {
+                state = (OnOffType) value;
+            } else {
+                logger.warn("Update channel {}: Unsupported value type {}", channelId,
+                        value.getClass().getSimpleName());
+            }
+            logger.debug("Update channel {} with state {} ({})", channelId, (state == null) ? "null" : state.toString(),
+                    (value == null) ? "null" : value.getClass().getSimpleName());
+
+            // Update the channel
+            if (state != null) {
+                updateState(channelId, state);
+            }
+        }
+    }
+
+    /**
+     * Build request URL from configuration data
+     *
+     * @return a valid URL for the thermostat's JSON interface
+     */
+    private String buildRequestURL(String resource) {
+        RadioThermostatConfiguration config = getConfigAs(RadioThermostatConfiguration.class);
+
+        String hostName = config.hostName;
+
+        String urlStr = URL.replace("%hostName%", hostName);
+        urlStr = urlStr.replace("%resource%", resource);
+
+        return urlStr;
+    }
+
+    /**
+     * Request new data from the thermostat
+     *
+     * @param the resource URL constant for a particular thermostat JSON resource
+     * @return an object mapping to one of the various thermostat JSON responses or null in case of error
+     */
+    private @Nullable Object getRadioThermostatData(String resource) {
+        Object result = null;
+        String errorMsg = null;
+
+        String urlStr = buildRequestURL(resource);
+        logger.debug("URL = {}", urlStr);
+
+        try {
+            // Run the HTTP request and get the JSON response from the thermostat
+            ContentResponse contentResponse = httpClient.newRequest(urlStr).method(GET).timeout(20, TimeUnit.SECONDS)
+                    .send();
```

Comment:
@cpmeister is it ok to use a *blocking* HTTP request on a scheduler thread?

Reply:
A blocking request here is fine.

Reply:
The only time that you wouldn't want to use the scheduler thread is when you are running a task that could take longer than a couple of seconds. In that case a dedicated thread/threadpool would be more appropriate.

Reply:
If I see correctly, the thread is blocked for 20 sec. if the device is not reachable. Is that acceptable?

Reply:
Couldn't you set a timeout for the request so that it would only block for a couple of seconds at most?

Reply:
I think an http timeout of 20 seconds is a little unreasonable. Even in the worst conditions a timeout of 3 seconds should be more than sufficient.

Reply:
Ordinarily I would agree. But this thing is slow as molasses. It can only process one request at a time. And if it is busy doing other things, the response will take a while. But if you got past the point of the device responding with the SSDP packet, there is not any issue with reachability. The 20 seconds was for insurance because I saw it timeout frequently with the timeout set for 10 seconds.

Reply:
What about using Jettys async API?
```
httpClient.newRequest("http://domain.com/path")
        .send(result -> { /* Your logic here */ });
```

Reply:
If it is going to be a problem I can get rid of the Http calls in the discovery. The IP address would be enough to create the thing. Getting additional info via Http was needed to provide the friendly name of the thermostat to the inbox.

Reply:
> 
> 
> What about using Jetty's async API?
> 
I though we had already gone down that road but the issues injecting openhab's httpClient instance into the discovery service was too much overhead.

@cpmeister

Reply:
I think discovery runs in a dedicated thread, which is not a problem. But the code we're talking about is in the handler. I just saw that this HTTP request is even invoked within `initialize()`, which should return fast.

Reply:
Yep. initialize() just calls the scheduler now.

Reply:
For the RT I wait 3 seconds, and retry 3 times before giving up. I haven't had any permanent timeouts in a month (I have 4 of the CT50 model).

```
    private String getData(String path) throws RadioThermostatCommunicationException {
        String content;
        for (int i = 0; i < RETRY_ATTEMPTS; i++) {
            try {
                String requestString = baseURL + path;
                logger.trace("Requesting {}", requestString);

                ContentResponse response = httpClient.newRequest(requestString)
                        .timeout(TIMEOUT_SECONDS, TimeUnit.SECONDS).send();
                logger.trace("Response code {}", response.getStatus());
                if (response.getStatus() != 200) {
                    throw new RadioThermostatCommunicationException(
                            "Error communicating with thermostat. Error Code: " + response.getStatus());
                }
                content = response.getContentAsString();
                logger.trace("sendRequest: response {}", content);
                return content;
            } catch (InterruptedException | ExecutionException e) {
                throw new RadioThermostatCommunicationException(e);
            } catch (TimeoutException e) {
                logger.debug("Timeout", e);
            }
        }
        throw new RadioThermostatCommunicationException("All retries timed out.");
    }
```

Reply:
Ok I'll change the timeout in the handler to 3 seconds and do some testing with it tonight. If I don't have issues with 3 second timeouts I will leave it synchronous. Otherwise I will do async as suggested by @fwolter.

Reply:
Using the async api is a completely viable alternative if you want to go that route.

Reply:
@billfor I have a CT-80 Rev A that I use as for testing. The connection failed the first time when I sent the timeout to 3 seconds. I think I might need to re-think the design of the binding to deal with async. Right now the getRadioThermostatData() calls expect an object to be returned by the method. I think it would work better for async if the calls send an event when they finally finish and then an event listener will receive the event and update the channels accordingly.

Reply:
It does time out a lot, but most of the time after three retries it seems to work for me.  I poll it at a frequency of 60 seconds, so it might always miss one entirely but pick up another and I don't notice. I didn't see anything in my logs.